ospf6d: add support for NSSA totally stub areas

rwestphal opened the pull request:
Add a knob to turn an NSSA area into a totally stub area. In this configuration, a Type-3 default summary route is generated by default.

Syntax: `area A.B.C.D nssa no-summary`.

Example:
```
router ospf6
 area A.B.C.D nssa no-summary
!
```

Note: different from stub areas, NSSA areas can have Type-7 external routes redistributed in from another routing domain. And among those external routes there might be a default route, which can conflict with the ABR default route if one exists. For this reason, the NSSA RFC specifies that an NSSA ABR should advertise a Type-3 default route **only** when summary routes aren't imported into the NSSA (i.e. totally stub NSSA). _ospfd_ violates that part of the spec since an NSSA ABR announces a Type-3 default route regardless if the "no-summary" option is configured or not. That gives operators little leeway to manipulate default routes according to business needs. In this PR I decided to make ospf6d comply to the NSSA specification instead of mirroring ospfd's behavior, which probably needs to be fixed at some point.

polychaeta reviewed (commented):
Thanks for your contribution to FRR!

Style checking failed; check logs

If you are a new contributor to FRR, please see our [contributing guidelines](http://docs.frrouting.org/projects/dev-guide/en/latest/workflow.html#coding-practices-style).

After making changes, you do not need to create a new PR. You should perform an [amend or interactive rebase](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) followed by a [force push](https://git-scm.com/docs/git-push#Documentation/git-push.txt---force).

polychaeta reviewed (commented):
Thanks for your contribution to FRR!


<details>
<summary><b>Click for style suggestions</b></summary>

To apply these suggestions:

<p>

```
curl -s https://gist.githubusercontent.com/polychaeta/99fbe8df6b00411d7406c0ec775af935/raw/af151be239af0ed31244b4d7d8c3acd513a83b95/cr_9491_1630386551.diff | git apply
```

</p>

<p>

```diff
diff --git a/ospf6d/ospf6_asbr.c b/ospf6d/ospf6_asbr.c
index 5b272b8f3..b5bf81c21 100644
--- a/ospf6d/ospf6_asbr.c
+++ b/ospf6d/ospf6_asbr.c
@@ -610,8 +610,7 @@ void ospf6_asbr_lsa_add(struct ospf6_lsa *lsa)
 			   OSPF6_PREFIX_OPTION_P)
 		|| oa->no_summary)) {
 		if (IS_OSPF6_DEBUG_EXAMIN(AS_EXTERNAL))
-			zlog_debug(
-				"Skipping Type-7 default route");
+			zlog_debug("Skipping Type-7 default route");
 		return;
 	}
 
diff --git a/tests/topotests/ospf6_topo2/test_ospf6_topo2.py b/tests/topotests/ospf6_topo2/test_ospf6_topo2.py
index bf492a61a..83e67bfe3 100644
--- a/tests/topotests/ospf6_topo2/test_ospf6_topo2.py
+++ b/tests/topotests/ospf6_topo2/test_ospf6_topo2.py
@@ -280,11 +280,13 @@ def test_redistribute_metrics():
 
     route = {
         "2001:db8:500::/64": {
-            "metricType":2, 
-            "metricCost":10,
+            "metricType": 2,
+            "metricCost": 10,
         }
     }
-    logger.info("Expecting AS-external route 2001:db8:500::/64 to show up with default metrics")
+    logger.info(
+        "Expecting AS-external route 2001:db8:500::/64 to show up with default metrics"
+    )
     expect_ospfv3_routes("r2", route, wait=30, detail=True)
 
     # Change the metric of redistributed routes of the static type on r3.
@@ -298,15 +300,16 @@ def test_redistribute_metrics():
     # Check if r3 reinstalled 2001:db8:500::/64 using the new metric type and value.
     route = {
         "2001:db8:500::/64": {
-            "metricType":1, 
-            "metricCost":60,
+            "metricType": 1,
+            "metricCost": 60,
         }
     }
-    logger.info("Expecting AS-external route 2001:db8:500::/64 to show up with updated metric type and value")
+    logger.info(
+        "Expecting AS-external route 2001:db8:500::/64 to show up with updated metric type and value"
+    )
     expect_ospfv3_routes("r2", route, wait=30, detail=True)
 
 
-
 def test_nssa_lsa_type7():
     """
     Test that static route gets announced as external route when redistributed
@@ -336,9 +339,7 @@ def test_nssa_lsa_type7():
     route = {
         "2001:db8:100::/64": {
             "pathType": "E1",
-            "nextHops": [
-                {"nextHop": "::", "interfaceName": "r4-eth0"}
-            ]
+            "nextHops": [{"nextHop": "::", "interfaceName": "r4-eth0"}],
         }
     }
 
@@ -357,14 +358,15 @@ def test_nssa_lsa_type7():
 
     def dont_expect_lsa(unexpected_lsa):
         "Specialized test function to expect LSA go missing"
-        output = tgen.gears["r4"].vtysh_cmd("show ipv6 ospf6 database type-7 detail json", isjson=True)
-        for lsa in output['areaScopedLinkStateDb'][0]['lsa']:
+        output = tgen.gears["r4"].vtysh_cmd(
+            "show ipv6 ospf6 database type-7 detail json", isjson=True
+        )
+        for lsa in output["areaScopedLinkStateDb"][0]["lsa"]:
             if lsa["prefix"] == unexpected_lsa["prefix"]:
                 if lsa["forwardingAddress"] == unexpected_lsa["forwardingAddress"]:
                     return lsa
         return None
 
-
     logger.info("Expecting LSA type-7 and OSPFv3 route 2001:db8:100::/64 to go away")
 
     # Test that LSA doesn't exist.

```

</p>
</details>


If you are a new contributor to FRR, please see our [contributing guidelines](http://docs.frrouting.org/projects/dev-guide/en/latest/workflow.html#coding-practices-style).

After making changes, you do not need to create a new PR. You should perform an [amend or interactive rebase](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) followed by a [force push](https://git-scm.com/docs/git-push#Documentation/git-push.txt---force).

polychaeta reviewed (commented):
Thanks for your contribution to FRR!


<details>
<summary><b>Click for style suggestions</b></summary>

To apply these suggestions:

<p>

```
curl -s https://gist.githubusercontent.com/polychaeta/457edab31895773dbb6bc6fe6ac939d8/raw/3942127ed4bbd93e4188eda12e6c6e700e38762d/cr_9491_1630386893.diff | git apply
```

</p>

<p>

```diff
diff --git a/tests/topotests/ospf6_topo2/test_ospf6_topo2.py b/tests/topotests/ospf6_topo2/test_ospf6_topo2.py
index a8642010c..83e67bfe3 100644
--- a/tests/topotests/ospf6_topo2/test_ospf6_topo2.py
+++ b/tests/topotests/ospf6_topo2/test_ospf6_topo2.py
@@ -280,11 +280,13 @@ def test_redistribute_metrics():
 
     route = {
         "2001:db8:500::/64": {
-            "metricType":2, 
-            "metricCost":10,
+            "metricType": 2,
+            "metricCost": 10,
         }
     }
-    logger.info("Expecting AS-external route 2001:db8:500::/64 to show up with default metrics")
+    logger.info(
+        "Expecting AS-external route 2001:db8:500::/64 to show up with default metrics"
+    )
     expect_ospfv3_routes("r2", route, wait=30, detail=True)
 
     # Change the metric of redistributed routes of the static type on r3.
@@ -298,11 +300,13 @@ def test_redistribute_metrics():
     # Check if r3 reinstalled 2001:db8:500::/64 using the new metric type and value.
     route = {
         "2001:db8:500::/64": {
-            "metricType":1, 
-            "metricCost":60,
+            "metricType": 1,
+            "metricCost": 60,
         }
     }
-    logger.info("Expecting AS-external route 2001:db8:500::/64 to show up with updated metric type and value")
+    logger.info(
+        "Expecting AS-external route 2001:db8:500::/64 to show up with updated metric type and value"
+    )
     expect_ospfv3_routes("r2", route, wait=30, detail=True)
 
 

```

</p>
</details>


If you are a new contributor to FRR, please see our [contributing guidelines](http://docs.frrouting.org/projects/dev-guide/en/latest/workflow.html#coding-practices-style).

After making changes, you do not need to create a new PR. You should perform an [amend or interactive rebase](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) followed by a [force push](https://git-scm.com/docs/git-push#Documentation/git-push.txt---force).

rwestphal commented:
> code looks good ... needs a documentation change, I think (?)

@riw777 docs updated. Let me know what you think.

I also rebased the code to fix a merge conflict and fixed the style issues relevant to changes that were done.

riw777 reviewed (approved):
looks good

riw777 commented:
ci:rerun

rwestphal commented:
CI:rerun
